### PR TITLE
Wip/docker refactoring

### PR DESCRIPTION
**Pull request by muuki88**
First rework of the docker API according to #453 

Reviews appreciated @fiadliel , @huntc , @hseeberger

### CODE REVIEW

File: src/main/scala/com/typesafe/sbt/packager/docker/DockerPlugin.scala [hunk 1]
```diff
@@ -91,79 +113,128 @@ object DockerPlugin extends AutoPlugin {
...
-      else
-        Some(Cmd("MAINTAINER", maintainer))
-    }
-
+  /**
+   * @param maintainer (optional)
+   * @return MAINTAINER if defined
+   */
+  private final def makeMaintainer(maintainer: String): Option[CmdLike] =
+    if (maintainer == null || maintainer.isEmpty) None else Some(Cmd("MAINTAINER", maintainer))
```
Comment (huntc) [inline src/main/scala/com/typesafe/sbt/packager/docker/DockerPlugin.scala]: How can maintainer ever be null?

File: src/main/scala/com/typesafe/sbt/packager/docker/DockerPlugin.scala [hunk 2]
```diff
@@ -91,79 +113,128 @@ object DockerPlugin extends AutoPlugin {
...
+   *
+   * All directories created afterwards are chowned.
+   *
+   * @param exposedVolumes
+   * @return commands to create, chown and declare volumes
+   * @see http://stackoverflow.com/questions/23544282/what-is-the-best-way-to-manage-permissions-for-docker-shared-volumes
+   * @see https://docs.docker.com/userguide/dockervolumes/
+   */
+  private final def makeVolumes(exposedVolumes: Seq[String], daemonUser: String, daemonGroup: String): Seq[CmdLike] = {
+    if (exposedVolumes.isEmpty) Seq()
```
Comment (huntc) [inline src/main/scala/com/typesafe/sbt/packager/docker/DockerPlugin.scala]: `Seq.empty`

**Comment by huntc** [comment]
LGTM - just one use of null that concerns me.

File: src/main/scala/com/typesafe/sbt/packager/docker/DockerPlugin.scala (hunk 2 above)
Comment (hseeberger) [inline src/main/scala/com/typesafe/sbt/packager/docker/DockerPlugin.scala]: `Nil`?

File: src/main/scala/com/typesafe/sbt/packager/docker/DockerPlugin.scala (hunk 1 above)
Comment (muuki88) [inline src/main/scala/com/typesafe/sbt/packager/docker/DockerPlugin.scala]: Too much Java code. Should not be possible

**Comment by muuki88** [comment]
Thanks for taking a look. I removed the critical parts.

**Comment by huntc** [comment]
LGTM

**Comment by andypetrella** [comment]
Would be awesome to have this one merged, so that I could easily be one of your first user of these features.

In two words: _need it_
:-D

cheers,
andy

**Comment by muuki88** [comment]
Available in `1.0.0-M5`

**Pull request merged**